ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,190 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    UNICODE_STRING AliasKeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\");
+    UNICODE_STRING BasicInfoName, InLinkName;
+    UNICODE_STRING GuidString;
+    HANDLE InterfaceKey = NULL;
+    PKEY_FULL_INFORMATION FullInfo = NULL;
+    PKEY_BASIC_INFORMATION BasicInfo = NULL;
+    PWCHAR pdest;
+    ULONG i, Size;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    /* Sanity check */
+    ASSERT(SymbolicLinkName != NULL);
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    BaseKeyName.Length = 0;
+    BaseKeyName.MaximumLength += GuidString.Length;
+    BaseKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                               BaseKeyName.MaximumLength,
+                                               TAG_IO);
+    if (!BaseKeyName.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    /* Concatenate alias GUID string with base key */
+    Status = RtlAppendUnicodeStringToString(&BaseKeyName, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Open base registry key */
+    Status = IoOpenDeviceInterfaceRegistryKey(&BaseKeyName,
+                                              GENERIC_READ,
+                                              &InterfaceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IoOpenDeviceInterfaceRegistryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        NULL,
+                        0,
+                        &Size);
+    if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    FullInfo = ExAllocatePoolWithTag(PagedPool,
+                                     Size,
+                                     TAG_IO);
+    if (!FullInfo)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        FullInfo,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    for (i = 0; i < FullInfo->SubKeys; i++)
```

Comment:
It's unnecessary here to query `KeyFullInformation` just for the number of subkeys. Simply loop until `ZwEnumerateKey` returns `STATUS_NO_MORE_ENTRIES`. This both saves a bunch of code, and better handles the case where new subkeys are added while you're enumerating.

Reply:
OK, removed this. But if I'll loop for `Status != STATUS_NO_MORE_ENTRIES`, then I will probably need to move the first `ZwEnumerateKey` call outside the loop. I doubt: is it correct? Can we use endless `while (TRUE)`, which I added for now, similarly to as it done in `IoGetDeviceInterfaces`?

Reply:
Per the comment on `RtlCompareUnicodeString`, you probably don't need any loop or enumeration.
But yes, you can do a `while (TRUE)` with a `break`, or you could move the first `ZwEnumerateKey` call before the loop, or you could just set `Status = STATUS_SUCCESS` before the loop. All would solve this problem.